flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/StopWithSavepointContextTest.java

```diff
@@ -0,0 +1,249 @@
+        scheduler =
+                SchedulerTestingUtils.createSchedulerBuilder(
+                                jobGraph, ComponentMainThreadExecutorServiceAdapter.forMainThread())
+                        .setFutureExecutor(new DirectScheduledExecutorService())
+                        .build();
+        scheduler.startScheduling();
+
+        // the checkpoint scheduler is stopped before triggering the stop-with-savepoint
+        disableCheckpointScheduler();
+
+        testInstance = new StopWithSavepointContext(jobGraph.getJobID(), scheduler, this.log);
+    }
+
+    @Test
+    public void testHappyPathWithSavepointCreationBeforeTermination() throws Exception {
+        assertHappyPath(
+                (savepointPath) -> {
+                    testInstance.handleSavepointCreation(savepointPath, null);
+                    testInstance.handleExecutionTermination(
+                            Collections.singletonList(ExecutionState.FINISHED));
+                });
+    }
+
+    @Test
+    public void testHappyPathWithSavepointCreationAfterTermination() throws Exception {
+        assertHappyPath(
+                (savepointPath) -> {
+                    testInstance.handleExecutionTermination(
+                            Collections.singletonList(ExecutionState.FINISHED));
+                    testInstance.handleSavepointCreation(savepointPath, null);
+                });
+    }
+
+    private void assertHappyPath(Consumer<String> stopWithSavepointCompletion) throws Exception {
+        final String savepointPath = "savepoint-path";
+
+        stopWithSavepointCompletion.accept(savepointPath);
+
+        assertThat(testInstance.getResult().get(), is(savepointPath));
+
+        // the happy path won't restart the CheckpointCoordinator
+        assertCheckpointSchedulingDisabled();
+    }
+
+    @Test
+    public void testSavepointCreationFailure() {
+        final Exception exception = new Exception("Expected exception during savepoint creation.");
+        testInstance.handleSavepointCreation(null, exception);
+
+        try {
+            testInstance.getResult().get();
+            fail("An ExecutionException is expected.");
+        } catch (Throwable e) {
+            final Optional<Throwable> actualException =
+                    ExceptionUtils.findThrowableWithMessage(e, exception.getMessage());
+            assertTrue(actualException.isPresent());
+        }
+
+        assertCheckpointSchedulingEnabled();
+    }
+
+    @Test
+    public void testNoTerminationHandlingAfterSavepointCompletion() throws Exception {
+        assertNoTerminationHandling(
+                () -> {
+                    testInstance.handleSavepointCreation("savepoint-path", null);
+                    testInstance.handleExecutionTermination(
+                            // the task failed and was restarted
+                            Collections.singletonList(ExecutionState.RUNNING));
```

Comment:
`RUNNING` is not a terminal state. If an `Execution` is restarted, then the old `Execution` will reach a terminal state and then we will create a new `Execution`. Hence, we are testing something which is not possible atm I think.